django/db/models/query.py

```diff
@@ -327,9 +329,22 @@ def __or__(self, other):
             return other
         if isinstance(other, EmptyQuerySet):
             return self
-        combined = self._chain()
+
+        if not self.query.can_filter():
+            pks = Subquery(self.values('pk'))
```

Comment:
The [`Subquery` wrapping should not be necessary here](https://github.com/django/django/blob/024abe5b82d95ee60cb18a77ebf841ad715467fa/tests/queries/tests.py#L2057-L2069).